pyomo/contrib/sensitivity_toolbox/sens.py

```diff
@@ -7,216 +7,684 @@
+    if not kaug.available(False):
+        raise RuntimeError('k_aug is not available')
+    if not dotsens.available(False):
+        raise RuntimeError('dotsens is not available')
+
+    # Declare Suffixes
+    model.dual = Suffix(direction = Suffix.IMPORT)
+    model.ipopt_zL_out = Suffix(direction=Suffix.IMPORT)
+    model.ipopt_zU_out = Suffix(direction=Suffix.IMPORT)
+    model.ipopt_zL_in = Suffix(direction=Suffix.EXPORT)
+    model.ipopt_zU_in = Suffix(direction=Suffix.EXPORT)
 
-    #Verify User Inputs    
-    if len(paramSubList)!=len(perturbList):
-        raise ValueError("Length of paramSubList argument does not equal "
-                        "length of perturbList")
-
-    for pp in paramSubList:
-        if pp.ctype is not Param:
-            raise ValueError("paramSubList argument is expecting a list of Params")
-
-    for pp in paramSubList:
-        if not pp._mutable:
-            raise ValueError("parameters within paramSubList must be mutable")  
-
-        
-    for pp in perturbList:
-        if pp.ctype is not Param:
-            raise ValueError("perturbList argument is expecting a list of Params")
-    #Add model block to compartmentalize all sipopt data
-    b=Block()
-    block_name = unique_component_name(instance, '_sipopt_data')
-    instance.add_component(block_name, b)
-
-    #Based on user input clone model or use orignal model for anlaysis
-    if cloneModel:
-        b.tmp_lists = (paramSubList, perturbList)
-        m = instance.clone()
-        instance.del_component(block_name)
-        b = getattr(m, block_name)
-        paramSubList, perturbList = b.tmp_lists
-        del b.tmp_lists
+    # K_AUG SUFFIXES
+    model.dof_v = Suffix(direction=Suffix.EXPORT)  #: SUFFIX FOR K_AUG
+    model.rh_name = Suffix(direction=Suffix.IMPORT)  #: SUFFIX FOR K_AUG AS WELL
+    kaug.options["print_kkt"] = ""
+    results = ipopt.solve(model,tee=tee)
+
+    # Rasie Exception if ipopt fails 
+    if (results.solver.status == SolverStatus.warning):
+        raise Exception(results.solver.Message)
+
+    for o in model.component_objects(Objective, active=True):
+        f_mean = value(o)
+    model.ipopt_zL_in.update(model.ipopt_zL_out)
+    model.ipopt_zU_in.update(model.ipopt_zU_out)
+    #: run k_aug
+    kaug.solve(model, tee=tee)  #: always call k_aug AFTER ipopt.
+    model.write('col_row.nl', format='nl', io_options={'symbolic_solver_labels':True})
+    # get the column numbers of theta
+    line_dic = {}
+    try:
+        for v in theta_names:
+            line_dic[v] = line_num('col_row.col', v)
+        # load gradient of the objective function
+        gradient_f = np.loadtxt("./GJH/gradient_f_print.txt")
+        with open ("col_row.col", "r") as myfile:
+            col = myfile.read().splitlines()
+    except Exception as e:
+        print('File not found.')
+    gradient_f_dic = {}
+    for i in range(len(col)):
+        gradient_f_dic["d(f)/d("+col[i]+")"] = gradient_f[i]
```

Comment:
I would prefer if this was a dictionary mapping integer indices to values.

Reply:
I will discuss it with Alex.